### PR TITLE
feat(rpc): add partial `eth_getProof` implementation

### LINKED ISSUE #1507: Add `eth_getProof` impl
### Describe the feature

This endpoint returns the Merkle proof for a given account and optionally some storage keys

### Additional context

This endpoint should interact with the StateProvider to get the proof. For this, a new method is needed in the Merkle Patricia Tree implementation.

### PR DESCRIPTION
**Pull request by MegaRedHand**
Closes: https://github.com/paradigmxyz/reth/issues/1507

This PR adds the implementation for the `eth_getProof` RPC call, along with additional methods in `DBTrieLoader` for generating Merkle proofs.

### CODE REVIEW

**Comment by MegaRedHand** [comment]
@mattsse 
I'm having a bit of trouble with this. As it stands, the `HistoricalStateProvider` would need to unwind the hashed account and storage tables to the past state before updating the trie and generating the proof. All of this would be really expensive and needs to be done while leaving the current tables untouched. I was thinking of just doing this in the transaction without committing (so changes are dropped), but this wouldn't work as it is read-only. Do you know of something that could be used for this?

On another note, the `LatestStateProvider` doesn't seem to be used aside from on tests.

**Comment by joshieDo** [comment]
My impression is that we need to change the way we insert into the trie - and use a cursor instead (included that remark in the other PR). this would allow us to have more control over what we commit.

For this kind of unwinding, we have to "force"  some hashing stages to use their incremental branch instead of batched/scratch which don't commit the db tx iirc

**Comment by rakita** [comment]
There could be only one write tx, this could block other write transactions.

**Comment by MegaRedHand** [comment]
**Update:** I left the implementation for the latest state provider only. If the block requested isn't the latest, it returns a "not implemented" error via RPC. As part of this, I modified the state access helpers to allow returning any of both the latest and history state providers.

I've noticed that _geth_ uses multiple databases for caching previous states, and use this for the endpoint backend. Other implementations seem to have partial (only for the latest or cached) or no support for the endpoint.


> My impression is that we need to change the way we insert into the trie - and use a cursor instead

Because of limitations with _cita_trie_ and the need for cursor mutability, this would require a `Mutex<RefCell<_>>` inside the database wrapper. I can add this if it seems like a good enough tradeoff, but maybe we should first add the stage benchmarks and check it doesn't drop performance.

> For this kind of unwinding, we have to "force" some hashing stages to use their incremental branch instead of batched/scratch which don't commit the db tx iirc

That's right for the `execute` part, but the `unwind` doesn't `commit` changes. Also, the transaction given to the provider is read-only so we can't make changes to the DB and we can't make this RW because of what rakita said.

**Comment by joshieDo** [comment]
blocked by #1497

**Comment by gakonst** [comment]
Needs merge conflict post #1474, sorry about that.

File: crates/rpc/rpc/src/eth/api/mod.rs
Comment (gakonst) [inline crates/rpc/rpc/src/eth/api/mod.rs:105]: Let's please avoid really short names like this

File: crates/rpc/rpc/src/eth/api/server.rs
Comment (gakonst) [inline crates/rpc/rpc/src/eth/api/server.rs:384]:
```suggestion
            EthApiError::InvalidBlockRange => internal_rpc_err("eth_getProof is unimplemented for historical blocks"),
```

File: crates/rpc/rpc/src/eth/api/mod.rs
Comment (gakonst) [inline crates/rpc/rpc/src/eth/api/mod.rs:153]: Why is this change needed? `latest_state()` did the same before, no?

File: crates/rpc/rpc/src/eth/api/state.rs
Comment (gakonst) [inline crates/rpc/rpc/src/eth/api/state.rs:74]:
```suggestion
            let (account_proof, storage_hash, stg_proofs) = state.proof(address, &keys)?;
```

File: crates/storage/provider/src/providers/state/historical.rs
Comment (gakonst) [inline crates/storage/provider/src/providers/state/historical.rs:127]:
```suggestion
        _keys: &[H256],
```
can't do like this?

File: crates/storage/provider/src/providers/state/latest.rs
Comment (gakonst) [inline crates/storage/provider/src/providers/state/latest.rs:74]: let's use full words, `storage_proof`, `account_proof` etc.

File: crates/storage/provider/src/trie/mod.rs
Comment (gakonst) [inline crates/storage/provider/src/trie/mod.rs:264]: cc @joshieDo @onbjerg 

the reason why I don't like the Cita Trie lib is because it doesn't let us reuse our cursors when we're walking thru the entire table, feels a bit gross to re-instantiate a cursor on every `get`

File: crates/storage/provider/src/trie/mod.rs
Comment (gakonst) [inline crates/storage/provider/src/trie/mod.rs:502]: This is a weird API due to requiring the root to be passed by the user every time. Can we avoid that somehow?

File: crates/storage/provider/src/trie/mod.rs
Comment (gakonst) [inline crates/storage/provider/src/trie/mod.rs:525]: Vec<Vec<Vec<_>>> looks weird, can we introduce a new type for this?

File: crates/storage/provider/src/trie/mod.rs
Comment (gakonst) [inline crates/storage/provider/src/trie/mod.rs:824]: let's please add a test for something that has multiple storage slots as well

**Review by gakonst — CHANGES_REQUESTED** [review]
good work - mainly nits

File: crates/rpc/rpc/src/eth/api/mod.rs
Comment (MegaRedHand) [reply crates/rpc/rpc/src/eth/api/mod.rs:153]: Yes, but this way the consumer of the `Factory` doesn't need to explicitly use `LatestSP`, which would need an additional check so he knows if the `BlockId` received references the latest block of the chain or not. This check wasn't being done before, and we were always using the `HistorySP` by default.

File: crates/storage/provider/src/providers/state/historical.rs
Comment (MegaRedHand) [reply crates/storage/provider/src/providers/state/historical.rs:127]: Yes, you're right.

File: crates/storage/provider/src/trie/mod.rs
Comment (MegaRedHand) [reply crates/storage/provider/src/trie/mod.rs:528]: Maybe we can make the `DBTrieLoader` receive the current root on instantiation in a similar way to the database wrappers we use inside the module (i.e. with `new(root)` and `empty()` methods).

File: crates/storage/provider/src/trie/mod.rs
Comment (MegaRedHand) [reply crates/storage/provider/src/trie/mod.rs:852]: Added

File: crates/storage/provider/src/trie/mod.rs
Comment (MegaRedHand) [reply crates/storage/provider/src/trie/mod.rs]: Added a type alias for `Vec<Vec<u8>>` named `MerkleProof`

File: crates/storage/provider/src/trie/mod.rs
Comment (gakonst) [reply crates/storage/provider/src/trie/mod.rs:532]: Yeah I like that.

**Review by gakonst — APPROVED** [review]
Good from me - would like @joshieDo to approve before merging, cc @mattsse given proximity to RPC-related code

File: crates/rpc/rpc/src/eth/api/state.rs
Comment (MegaRedHand) [reply crates/rpc/rpc/src/eth/api/state.rs:71]: No. It returns empty _storage_ proofs when the account doesn't exist. I misread the docs 😅

File: crates/storage/provider/src/trie/mod.rs
Comment (MegaRedHand) [reply crates/storage/provider/src/trie/mod.rs:532]: Opened https://github.com/paradigmxyz/reth/issues/1743. Should be a small change, and a good first issue.

**Review by gakonst — APPROVED** [review]
OK - seems fine at this point. Let's merge and we can iterate if needed.

**Pull request merged**
